### ISSUE TITLE
Failed to extract the current "stop releases"

### ISSUE THREAD

**Issue by ghost** [Olf0/sfos-upgrade#48]
Hey,
Is this warning dangerous for the upgrade operation? What could I do about it?

```
# sfos-upgrade 3.4.0.24
[D] unknown:0 - Got http_proxy from environment, will not talk to connman
Warning: Failed to extract the current "stop releases" from https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1
Hence using an internal, potentially outdated list of stop releases instead.
Do you really want to continue? (Y/N)
```
Also, why those information is needed?
Any help would be appreciated.

**Comment by Olf0** [comment]
> Is this warning dangerous for the upgrade operation?
No, because currently [sfos-upgrade's internal list of stop releases](https://github.com/Olf0/sfos-upgrade/blob/master/usr/bin/sfos-upgrade#L183) is [up-to-date](https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1) (plus 2.0.5.6 and 1.0.0.5).
Additionally downloading and extracting the current stop releases from https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1 is a security measure, as the internal list may be outdated.

> What could I do about it?
The error message looks as if a system-wide HTTP-Proxy is set on your device, which prevents the page https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1 from being downloaded. Does that web-page load in the native Browser on this device?

> Also, why those information is needed?
See https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1
TL;DR: "Jumping over" a stop release will likely break your SailfishOS installation.

**Comment by wickedsp1d3r** [comment]
Yes, I am using a HTTP proxy and have relevant environment variables set up.

> Does that web-page load in the native Browser on this device?

Yes.

But I noticed a request with cURL returns `403` from that URL:
```
$ curl -I https://jolla.zendesk.com/hc/en-us/articles/201836347
HTTP/1.0 200 Connection established
Proxy-agent: tinyproxy/1.10.0

HTTP/1.1 403 Forbidden
Date: Wed, 14 Oct 2020 17:39:52 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
...other headers...
```
Ignoring proxy with flag `--noproxy "*"` gives the same result.

**Comment by Olf0** [comment]
Well, then something is strange about your HTTP-proxy setup.
Interestingly, the [cURL call before](https://github.com/Olf0/sfos-upgrade/blob/master/usr/bin/sfos-upgrade#L163) that one seems to succeed with your setup.  You may check that at the command line per 
`curl -LO https://coderus.openrepos.net/whitesoft/sailversion`
If my assumption is correct, you have a positive and a negative test to experiment with your proxy settings.  

But as the only relevant difference between those two cURL calls is the DNS-domain visited (AFAICS), can you assure that no DNS-based filtering is involved (be it "on device", in your LAN, by your internet provider etc.)?

**Comment by carmenfdezb** [comment]
Hi @Olf0! I obtained the same warning when I updated my XA2 and my Jolla Tablet and I don't use proxy

**Comment by wickedsp1d3r** [comment]
It seems like a blank page with an error _"Please turn JavaScript on and reload the page."_ is returned from jolla.zendesk.com when `User-Agent` is `curl/<curl_version>`.

**Comment by Olf0** [comment]
Oh, now I see.  Thank you @wickedsp1d3r (and @carmenfdezb for confirming) for insisting on this.
I should have looked inside the file, when I tested downloading manually.  Yes, it does not have the expected content, which is displayed in the browser.  This was definitely working in spring 2020 (when SFOS 3.3.0 was released).

I am not sure, how this is related to @wickedsp1d3r's original proxy warning, but I will experiment with user agent settings etc. to let sfos-upgrade download and extract the recent list of stop releases from there, again.
Do you have any specific suggestions how to implement this?

**Comment by Olf0** [comment]
After some research and experimenting, I have the impression, that one has to enable JavaScript for Cloudflare (the content delivering company) to be able to see any content at jolla.zendesk.com

As I am not a web expert and have no idea how to resolve this (or even if my quick analysis is correct), I will likely delete or disable this security measure.  This is unfortunate, as sfos-upgrade cannot automatically determine the recent stop releases, hence increasing maintenance burden and responsibility. :-(

**Comment by Olf0** [comment]
Well, so I released [sfos-upgrade 3.7.8.](https://github.com/Olf0/sfos-upgrade/releases/tag/3.7.8) in which downloading the list of current stop releases from jolla.zendesk.com is disabled.

I still leave this issue open, in the hope someone has an idea how to circumvent the hurdles Zendesk / Jolla impose for accessing this list in an automated manner.

**Comment by carmenfdezb** [comment]
Thank you so much!

**Comment by Olf0** [comment]
Filing a request at [jolla.zendesk.com](https://jolla.zendesk.com/hc/en-us/requests/new) may help, which kindly asks to offer a current list of "SailfishOS stop releases" suitable for *automatic* ("scripted") download (i.e., per `curl`) either at [its usual place](https://jolla.zendesk.com/hc/en-us/articles/201836347#4.1) or another (then, also maybe in a better machine-readable form).

**Comment by Olf0** [comment]
Because none of the categories the bug reporting process at [jolla.zendesk.com](https://jolla.zendesk.com/hc/en-us/requests/new) offers is well suited for this, I decided to report this at FSO in the "Platform Development" category: https://forum.sailfishos.org/t/scripted-download-of-https-jolla-zendesk-com-hc-en-us-articles-201836347-fails/7935

**Comment by Olf0** [comment]
Accidentally I noticed that Jolla has silently provided a machine-downloadable list of "stop releases" they maintain, see [FSO issue 7935 "Scripted download of https://jolla.zendesk.com/hc/en-us/articles/201836347 fails!"](https://forum.sailfishos.org/t/scripted-download-of-https-jolla-zendesk-com-hc-en-us-articles-201836347-fails/7935/15?u=olf).

So finally I was able to fix this by [commit 5ed9547](https://github.com/Olf0/sfos-upgrade/commit/5ed9547db18719477e34689a221af54f33b74b3a).

**Comment by Olf0** [comment]
> After some research and experimenting, I have the impression, ~that one has to enable JavaScript for Cloudflare (the content delivering company) to be able to see any content at jolla.zendesk.com~
> 
> _Edit:_ No, it loads in Firefox with JavaScript disabled by NoScript, but setting the user agent string does not seem to make a difference, as `curl -A 'Mozilla/5.0 (X11; Linux x86_64) Gecko/20100101 Firefox/47' https://jolla.zendesk.com/hc/en-us/articles/201836347` still downloads a "No JavaScript"-warning page.
> 
> As I am not a web expert and have no idea how to resolve this (or even if my quick analysis is correct), […]. :-(

For some technical background on the bot-detection and -prevention measures the two big CDNs (Content Delivery Networks) Coudflare and Akamai emply, see [Forum.Sailfishos.Org: Scripted download of https://jolla.zendesk.com/hc/en-us/articles/201836347 fails! - Reply 16](https://forum.sailfishos.org/t/scripted-download-of-https-jolla-zendesk-com-hc-en-us-articles-201836347-fails/7935/16?u=olf).
From there:
<sub>[@flyping provided a pointer to the corresponding "bots"-related measures](https://forum.sailfishos.org/t/browser-is-blocked-by-akamai-and-cloudflare/12540/34?u=olf) the CDNs (Content Delivery Networks) [Cloudflare](https://developers.cloudflare.com/bots/concepts/bot-score/) and [Akamai](https://techdocs.akamai.com/cloud-security/docs/detection-methods) employ.
The [documentation by Cloudflare](https://developers.cloudflare.com/bots/concepts/bot-score/) describes very well, [what I experienced for almost 3 years](https://github.com/Olf0/sfos-upgrade/issues/48#issuecomment-709501852).  As expected these measures are designed to be non-circumventable.<br />
Side note: Both, [Akamai](https://techdocs.akamai.com/cloud-security/docs/about-bots) and [Cloudflare](https://developers.cloudflare.com/bots/plans/) emphasise, that the bot-related measures are highly customer-configurable and can be enabled per web-page (at least for Akamai).
Nevertheless, moving Jolla's official list of stop releases to the SailfishOS-Wiki and open it up for collaboration at GitHub is a vastly better solution.
Thanks!</sub>
